### DISCUSSION TITLE
Middleware not working with pinia stores (composition api)

### DISCUSSION THREAD

**Discussion by CODE-AXION** [nuxt/nuxt#24276] [Questions]
Hello guys i am stuck on a very simple problem , Which i think you guys could solve it in a minute . so please please someone help me 

i am trying to make a very simple middleware which will check if the user is already logged in dont let him visit the log in page thats it 

but whenever i am tryna use this middleware it just doesn't work, i have already wasted 6 hours on this straight.

when i use the ``` let checkUser = authStore.isLoggedIn ``` the value is always true for logged in users which i have debugged inside console.log but when i hard code the value of the checkuser variable then it works perfectly i dont know what is the problem here

here's what i am getting when i am using stores : the middleware waits for a sec and then redirects and gives me some errors which look like this :

![image](https://github.com/nuxt/nuxt/assets/97381867/b5545a15-aa22-411d-a200-3b3b6ae36f53)

also when i try to do the inverse it works i dont know how but i dont want that , you can see at the below code and the comment

```javascript
import { useAuth } from "../composables/Auth.js"

export default defineNuxtRouteMiddleware(async(to,from) => {
    const  nuxtApp  = useNuxtApp()
    const authStore = useAuth(nuxtApp.$pinia);
    let checkUser = ref(authStore.isLoggedIn);

    console.log(checkUser.value)
    
    // this doesn't work
    if ((to !== '/') && (checkUser.value)) {
   
        return navigateTo('/')
    }
    
    // this works somehow with the inverse condition but i dont want that 
    if ((to !== '/') && !(checkUser.value)) {
   
        return navigateTo('/')
    }

    // if (to !== '/' && !authStore.user?.id) return navigateTo("/", { replace: true });
});
```

**Comment by TheAlexLichter** [comment]
What is in your `Auth.js` file?

**Reply by CODE-AXION** [reply]
thank you so much man for noticing me , i am literally so bad stuck at this 

here i am using larafetch which i used from the breeze-nuxt package to add laravel authentication
https://github.com/amrnn90/breeze-nuxt/tree/master

in here everything works perfectly here there is no errors here, the problem is the middleware 


composables/Auth.js

```javascript
import { defineStore } from "pinia";


  
export const useAuth = defineStore("Auth", ()=>{
    
    const route = useRoute()

    const router = useRouter();
    const user = ref(null);
    const user_id = ref(null);
    
    const isLoggedIn = computed(() => !!user.value);

    async function login(credentials) {
        if (isLoggedIn.value) return;

        await $larafetch("/login", { method: "post", body: credentials,});

        await refresh();
    }

    async function register(credentials) {
        await $larafetch("/register", { method: "post", body: credentials });
        await refresh();
    }
      
    async function refresh() {
        try {
            
            let fetchedUser = await fetchCurrentUser();
            user_id.value = fetchedUser.id
            return user.value = fetchedUser;

            // console.log(user.value)
        } catch {
            
            user_id.value = null
            return user.value = null;
        }
    }

    async function setUser(setUser)
    {
        user.value = setUser;
    }

    async function logout() {
        if (!isLoggedIn.value) return;

        await $larafetch("/logout", { method: "post" });
        user.value = null;

        await router.push("/auth/login");
    }

    const AuthUser = async () => {
        return user.value;
    }

    const fetchCurrentUser = async () => {
        try {
            // const { data } = await useLarafetch("/api/user");
            // console.log(data)
          return await $larafetch("/api/user");
        } catch (error) {
          if ([401, 419].includes(error?.response?.status)) return null;
            throw error;
        }
    };

    return {
        user,
        isLoggedIn,
        user_id,
        fetchCurrentUser,
        login,
        setUser,
        register,
        logout,
        refresh,
    }

})
```

**Reply by TheAlexLichter** [reply]
Ohh, so `auth` is your store. I'd suggest naming it `useAuthStore` and put it in `/stores` then.

**Reply by CODE-AXION** [reply]
oh got it let me try it and then try to use middleware and see if it works or not !

**Reply by TheAlexLichter** [reply]
I wonder though, what does `checkedUser.value` return?

**Reply by TheAlexLichter** [reply]
checkedUser also should be a computed: `const checkedUser = computed(() => authStore.isLoggedIn)` or simpler - `const checkedUser = authStore.isLoggedIn`

**Comment by CODE-AXION** [comment]
> Ohh, so `auth` is your store. I'd suggest naming it `useAuthStore` and put it in `/stores` then.



> I wonder though, what does `checkedUser.value` return?
@manniL 

it returns if the user is logged in or not . if it is logged in it will return true else will return false

**Reply by TheAlexLichter** [reply]
Yeah, I got that :D
But what does it return in the middleware? Does it return true/false? Then the store works fine 👍🏻

Also see my remark above.

**Reply by CODE-AXION** [reply]
yes exactly it works what i am expecting , it returns true when user is logged in and false for vice versa

**Reply by TheAlexLichter** [reply]
What exactly is the problem then? 🤔

**Reply by CODE-AXION** [reply]
@manniL 

i seriously wanna cry out so loud , even after shifting the stores to its own folder its still giving me the same error, i couldn't even found related answer anywhere on the internet. 

![image](https://github.com/nuxt/nuxt/assets/97381867/eba5e91e-718b-4c02-b7bd-e010d170cbff)


![image](https://github.com/nuxt/nuxt/assets/97381867/8c3786bf-d207-4b14-9815-fcae3e69a80e)

**Reply by CODE-AXION** [reply]
@manniL 

everything works perfectly except the middleware part, the plugins , components, stores everything is working none of them is giving the errors .

**Reply by TheAlexLichter** [reply]
Okay, two more things: 

1. Do you use SSR?
2. Do you save user info in localstorage?

**Reply by CODE-AXION** [reply]
> Okay, two more things: 
> 
> 1. Do you use SSR?
> 2. Do you save user info in localstorage?

@manniL 

Yes I use SSR

And no I don't save user info in localstoage

**Reply by TheAlexLichter** [reply]
How do you store the user data instead?

**Comment by CODE-AXION** [comment]
> How do you store the user data instead?

In Sessions. laravel api authentication using sessions 

the api is statefull here

**Reply by CODE-AXION** [reply]
@manniL 

brother if possible i would create a reproduction , would you check why this middleware not workin ?

i really want to solve this problem , my whole development is stucked because of this , i saw other youtube videos and i am doing the same thing , but mine is not working and their's is working ...

**Reply by TheAlexLichter** [reply]
Sure, if you create a reproduction I'll have a look 👍

**Comment by joaopalopes24** [comment]


**Reply by CODE-AXION** [reply]
> @CODE-AXION Did you manage to solve this problem?

Yess , was running on different ip and port

I did php artisan serve --host localhost --port=8000

After this it worked

Previously I was just doin php artisan serve that's why
